consensus/src/data_manager.rs

```diff
@@ -0,0 +1,234 @@
+// Copyright (c) Aptos
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::quorum_store::utils::RoundExpirations;
+use aptos_crypto::HashValue;
+use aptos_infallible::Mutex;
+use aptos_logger::{debug, warn};
+use aptos_types::transaction::SignedTransaction;
+use arc_swap::ArcSwapOption;
+use consensus_types::{
+    block::Block,
+    common::Payload,
+    proof_of_store::{LogicalTime, ProofOfStore},
+    request_response::WrapperCommand,
+};
+use dashmap::DashMap;
+use executor_types::*;
+use futures::channel::mpsc::Sender;
+use std::sync::Arc;
+use std::sync::atomic::{AtomicBool, Ordering};
+use tokio::sync::oneshot;
+use crate::quorum_store::batch_reader::BatchReader;
+
+
+enum DataStatus {
+    Cached(Vec<SignedTransaction>),
+    Requested(Vec<oneshot::Receiver<Result<Vec<SignedTransaction>, Error>>>),
+}
+
+/// Execution -> QuorumStore notification of commits.
+pub struct DataManager {
+    quorum_store_enabled: AtomicBool,
+    data_reader: ArcSwapOption<BatchReader>,
+    quorum_store_wrapper_tx: ArcSwapOption<Sender<WrapperCommand>>,
+    digest_status: DashMap<HashValue, DataStatus>,
+    expiration_status: Mutex<RoundExpirations<HashValue>>,
+}
+
+impl DataManager {
+    /// new
+    pub fn new() -> Self {
+        Self {
+            quorum_store_enabled: AtomicBool::new(false),
+            data_reader: ArcSwapOption::from(None),
+            quorum_store_wrapper_tx: ArcSwapOption::from(None),
+            digest_status: DashMap::new(),
+            expiration_status: Mutex::new(RoundExpirations::new()),
+        }
+    }
+
+    #[allow(dead_code)]
+    pub fn enable_quorum_store(&self) {
+        self.quorum_store_enabled.store(true, Ordering::Relaxed)
+    }
+
+    async fn request_data(
+        &self,
+        poss: Vec<ProofOfStore>,
+        logical_time: LogicalTime,
+    ) -> Vec<oneshot::Receiver<Result<Vec<SignedTransaction>, executor_types::Error>>> {
+        let mut receivers = Vec::new();
+        for pos in poss {
+            debug!(
+                "QSE: requesting pos {:?}, digest {}, time = {:?}",
+                pos,
+                pos.digest(),
+                logical_time
+            );
+            if logical_time <= pos.expiration() {
+                receivers.push(
+                    self.data_reader
+                        .load()
+                        .as_ref()
+                        .unwrap() //TODO: can this be None? Need to make sure we call new_epoch() first.
```

Comment:
Should we address the TODO? otherwise, when.

Reply:
Let's double-check with @zekun000. The question is whether state_computer can call commit() / execute() before start_round_manager is called in epoch_manager.